### PR TITLE
feat(embodied): add FastWAM world-action model (LIBERO eval + SFT)

### PR DESCRIPTION
**Pull request by wzii**
Wrap the upstream FastWAM (Wan2.2 video-diffusion world-action model) behind RLinf's embodied interfaces for LIBERO eval and SFT (predict_action_batch -> infer_action, sft_forward -> training_loss).

### Description
This PR integrates FastWAM into RLinf.

### How has this been tested?
Tested with LIBERO-spatial evaluation (SR meets the paper's results) and a smoke run of sft

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/fastwam/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# FastWAM in RLinf
```
Comment (F9rozen) [inline rlinf/models/embodiment/fastwam/README.md:1]: Refer to the official rlinf documentation format

File: evaluations/libero/libero_spatial_fastwam_eval.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+defaults:
```
Comment (F9rozen) [inline evaluations/libero/libero_spatial_fastwam_eval.yaml:1]: Merge these YAMLs

File: rlinf/data/datasets/fastwam/__init__.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
+# Copyright 2026 The RLinf Authors.
```
Comment (thereAreDemonsNearby) [inline rlinf/data/datasets/fastwam/__init__.py:1]: You have only one __init__.py file for this dataset. Why not only rlinf/data/datasets/fastwam.py ?

File: rlinf/models/embodiment/fastwam/fastwam_policy.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        """
+        cfg = self.infer_cfg
+        main_images = env_obs["main_images"]
+        wrist_images = env_obs.get("wrist_images")
+        states = env_obs["states"]
+        task_descriptions = env_obs["task_descriptions"]
+        batch_size = int(main_images.shape[0])
+
+        chunks = []
+        for i in range(batch_size):
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/fastwam_policy.py:185]: A sample-by-sample inference scheme is unacceptably slow—could you implement a batched inference approach?

File: rlinf/models/embodiment/fastwam/fastwam_policy.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        return out
+
+    def default_forward(self, **kwargs):
+        raise NotImplementedError(
+            "FastWAMPolicy.default_forward is unused; rollout uses predict_action_batch "
+            "and SFT uses sft_forward."
+        )
+
+    def gradient_checkpointing_enable(self, **kwargs):
+        # FastWAM toggles gradient checkpointing through its DiT configs
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/fastwam_policy.py:271]: hope we can control gradient checkpointing through RLinf FSDP manager. be consistent with other models.

File: rlinf/models/embodiment/fastwam/fastwam_policy.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        (round(src_w * scale), round(src_h * scale)), resample=Image.BILINEAR
+    )
+    rw, rh = resized.size
+    left = max((rw - width) // 2, 0)
+    top = max((rh - height) // 2, 0)
+    cropped = resized.crop((left, top, left + width, top + height))
+    return np.asarray(cropped, dtype=np.uint8)
+
+
+class FastWAMPolicy(nn.Module, BasePolicy):
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/fastwam_policy.py:74]: When supporting other models, we've all been using inheritance—only you've used composition. Could you align with our approach?

File: rlinf/models/embodiment/fastwam/README.md (hunk 1 above)
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/README.md:1]: RLinf has a dedicated docs directory—please move all your documentation there. Also, the documentation isn't detailed enough yet; please add an explanation of how fastwam's config and RLinf's config interact, as it looks quite complex."

File: rlinf/models/embodiment/fastwam/__init__.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+def _compose_fastwam_cfg(config_dir: str, config_name: str, overrides) -> DictConfig:
+    """Compose a FastWAM Hydra config without disturbing RLinf's own Hydra state."""
+    from hydra import compose, initialize_config_dir
+    from hydra.core.global_hydra import GlobalHydra
+
+    _register_fastwam_resolvers()
+    already = GlobalHydra.instance().is_initialized()
+    if already:
+        GlobalHydra.instance().clear()
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/__init__.py:100]: I think this is not safe. Later use of rlinf hydra config would fail. Could you find a way like pure OmegaConf ？

File: rlinf/models/embodiment/fastwam/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+| `__init__.py` | `get_model` — composes FastWAM's Hydra configs, builds model + processor, loads checkpoint |
+| `../../../data/datasets/fastwam/` | SFT dataloader wrapping FastWAM `RobotVideoDataset` |
+
+Registered in `rlinf/config.py` (`SupportedModel.FASTWAM`, `EMBODIED_MODEL`),
+`rlinf/models/__init__.py` (`_build_fastwam`), and dispatched in
+`rlinf/workers/sft/fsdp_vla_sft_worker.py`.
+
+## Prerequisites
+
+- Install FastWAM as a package (`pip install -e /path/to/FastWAM`) so `import fastwam` works.
```
Comment (F9rozen) [inline rlinf/models/embodiment/fastwam/README.md:23]: need requirements guide

File: rlinf/models/embodiment/fastwam/fastwam_policy.py (hunk 4 above)
Comment (F9rozen) [inline rlinf/models/embodiment/fastwam/fastwam_policy.py:185]: Serial processing is slow

File: rlinf/models/embodiment/fastwam/__init__.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    if torch_dtype is None:
+        torch_dtype = torch.bfloat16
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+
+    from hydra.utils import instantiate
+
+    # Instantiate the FastWAM model (resolves ${data...}/${model...} against fcfg root).
+    model = instantiate(fcfg.model, model_dtype=torch_dtype, device=device)
+
+    ckpt_path = cfg.get("checkpoint_path", None) or cfg.get("model_path", None)
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/__init__.py:167]: Other models in RLinf only uses 'model_path'.    'checkpoint_path' is not expected.

File: rlinf/models/embodiment/fastwam/__init__.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+from omegaconf import DictConfig, OmegaConf
+
+from rlinf.models.embodiment.fastwam.fastwam_policy import FastWAMPolicy
+from rlinf.utils.logging import get_logger
+
+logger = get_logger()
+
+
+@dataclass
+class FastWAMInferConfig:
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/__init__.py:40]: This class is named **InferConfig**. Don't SFT or RL need these parameters? Some of them look like they're universal.

File: rlinf/models/embodiment/fastwam/fastwam_policy.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        from fastwam.utils.video_io import save_mp4
+
+        out_dir = self.infer_cfg.future_video_dir
+        os.makedirs(out_dir, exist_ok=True)
+        path = os.path.join(out_dir, f"future_{self._video_saves:04d}.mp4")
+        save_mp4(frames, path, fps=8)
+        self._video_saves += 1
+
+    # ------------------------------------------------------------------ utils
+    @property
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/fastwam/fastwam_policy.py:114]: If use inheritance instead of composition, devce and torch_dtype will not be necessary any more

**Review by thereAreDemonsNearby — CHANGES_REQUESTED** [review]

File: examples/sft/config/model/fastwam.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
+# FastWAM SFT model config.
+#
+# For SFT we initialise the MoT experts from the released FastWAM checkpoint
+# (skip_dit_load_from_pretrain=true in sim_libero), so the 20GB Wan DiT and the
+# ActionDiT backbone preprocessing are *not* required. Only the MoT experts and
+# the proprio encoder are trained (see get_model `freeze_non_dit`).
+model_type: "fastwam"
+precision: bf16
```
Comment (thereAreDemonsNearby) [inline examples/sft/config/model/fastwam.yaml:8]: For SFT, model precision should be set to fp32 to keep optimizer states in fp32. Fwd & bwd will be automatically run in bf16 precision because of fsdp mixed precision.

**Comment by zoeyuchao** [comment]
@wzii Hi, thanks for your contribution! Could you address the comment above so we can merge this pr asap?

**Comment by wzii** [comment]
> @wzii Hi, thanks for your contribution! Could you address the comment above so we can merge this pr asap?

Hi, I can't reproduce the evaluation results reported by fastwam using their ckpt within RLinf. I believe there is some effort needed to fix that and I haven't got the time to do so.
